RavenDB-20184 : Sharding - Restore - Ensure we can restore to a given point

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20184

### Additional description

- refactor `RestorePointBase` and its subclasses + minor modifications to `/admin/restore/points` endpoint
- add `LastFileNameToRestore` property to `SingleShardRestoreSetting`,  which will be used to restore each shard up to a specific backup file
- tests

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

aviv86 committed:
RavenDB-20184 : add LastFileNameToRestore property to SingleShardRestoreSetting, in order to support restoring a sharded database to a given restore-point (can be a different point-in-time for each shard)

aviv86 committed:
RavenDB-20184 : tests

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestorePointsBase.cs:
```diff
@@ -95,7 +95,32 @@ public class ParsedBackupFolderName
+        {
+            foreach (var path in paths)
+            {
+                await FetchRestorePoints(path, assertLegacyBackups);
```
Does it make sense to make this in parallel?

ppekrol replied:
Not a hot path IMO, not worth it

karmeli87 reviewed (commented):
Please check failing tests

aviv86 committed:
RavenDB-20184 : fix failing test

aviv86 committed:
RavenDB-20184 : refactor RestorePointsBase and its subclasses, and allow to have multiple restore-points with same date time in the result

aviv86 committed:
RavenDB-20184 : fix errors after rebase + revert some changes in RestorePointsBase -
use a sorted list as in the original code - since we are now proxying the request to all shard nodes, there's probably no need to support multiple restore-points with same date

aviv86 committed:
RavenDB-20184 : fix compilation after rebase

aviv86 commented:
@kalczur will open a new PR based on this branch with a working UI
